Document release strategy for Instructlab CLI (#985)
instructlab/instructlab: docs/release-strategy.md

`instructlab` Python package built from the
<https://github.com/instructlab/instructlab> git repository.

## Versioning Scheme

Releases use a `X.Y.Z` numbering scheme.

X-stream release are for major releases. At this stage in the project a major release has not been cut and we expect each release (2-3 weeks) to be a new Y-stream.

Backwards compatibility will exist within a Y-stream for example within 0.1.8 and 0.1.9 and between Y-stream releases for update purposes. For example between 0.1.9 and 0.2.0.

Z-stream releases are meant for critical bug and documentation fixes. Backward compatibility is guaranteed between Z-stream releases. Z-stream releases are cut as maintainers see fit.

## Schedule

The project currently operates on a time-based release schedule.
The goal is to release a new Y-stream every 2-3 weeks though the time span remains flexible based on the discretion of the maintainers team.

The cadence for major releases starting from 1.0 and onward will be determined as the project matures.

A schedule will be updated in a markdown file on the <https://github.com/instructlab/instructlab> GitHub repository.

## Release Tracking

> Release planning is tracked via milestones in GitHub. Milestones on GitHub will exist for the next two releases.

Reviewer: ```suggestion
Release planning is tracked via [milestones](https://github.com/instructlab/instructlab/milestones) in GitHub. Milestones on GitHub will exist for the next two releases.
```

> Release planning is tracked via milestones in GitHub. Milestones on GitHub will exist for the next two releases.

Reviewer 2: Why two?

> PRs and issues associated with the next two milestones will be prioritized for review and merging into the main branch of instructlab.

Reviewer 3: ```suggestion
PRs and issues associated with the next two milestones will be prioritized for review and merging into the `main` branch of instructlab.
```

> PRs and issues associated with the next two milestones will be prioritized for review and merging into the main branch of instructlab.

Reviewer 2: ```suggestion
PRs and Issues
```

## Git Branches and Tags

Every `X.Y` release stream gets a new branch.

> Each release, `X.Y.Z`, exists as a tag named `vX.Y.Z` on the `release-vX.Y` branch.

Reviewer 2: IMHO, we should delete all branches except for `main` and `release-0.14` currently present - none of the existing release branches follow this scheme and it's confusing frankly

## Release Branch Maintenance

Maintenance efforts are primarily focused on the `main` branch.
Occasionally, critical bug fixes may be backported to the most recent
`X.Y` branch for a Z-stream bug-fix release. Bug fixes in general are not

> actively backported to release branches.

Reviewer: I found this section to be blurry. I would be in favor of a stronger commitment on whether we do backport or not.
Also bringing clarity on the stable branches we support is important. Let's say we support the last two stable versions and everything below is not maintained anymore.

Reviewer 2: I'd go further and say we start off _only_ supporting the latest `stable` branch - currently `0.14.0` (soon to be `0.14.1`) - I don't see any advantage to supporting more than one release at this stage of the project (early, fast dev cycle, etc)

> ## Release Mechanics

Reviewer 3: One topic not covered is identifying who is responsible for running all of these steps. Here's a proposal:
```suggestion
## Release Mechanics

Release mechanics are done by a Release Manager identified for that release. The Release Manager is a member of the CLI Maintainers team that has agreed to take on these responsibilities. The Release Manager can change on a per-release basis. The Release Manager for each release is identified in the Description of the Milestone used to plan and track that release on GitHub.
```

The following are the steps for how a release gets cut.

1. Maintainers determine a commit on the main branch that will serve as the basis for the next release.
    - For a Z-stream release skip this step.
2. For a Y-Stream release: create a new branch in the format `release-vX.Y`.
    - For a Z-Stream release skip this step.
3. Create a commit for the release.
    - The subject of the commit will be `vX.Y.Z release` and there will be no body in the commit.
4. Open up a PR with the commit for the release.
5. Maintainers Merge PR.

>     - The PR will require 2 maintainer approvals.

Reviewer 3: I don't think you can create a completely empty commit. Is the intent to put tagging of a release through a review process? I think it is, but let me know if I'm wrong ...

An alternative to having that review is to create the branch and then create a PR on that release branch that updates a ChangeLog file or something like that with a version and some changelog text (could just be a link to the github milestone? tbd ...)

Reviewer 2: the way I understand it was this:

so everytime we do a new Y-stream:
1. new branch off main, tag X.Y.0 and stable

new Z stream:
1. backport commits to existing release branch
3. new tag for Z stream off latest commit to release branch
4. move stable tag to the same commit (should match new tag in step 2)

Reviewer 2: @cdoern and I will be doing a Z-stream release today, I will update here based on our experience

Reviewer 2: This is what we did for `0.14.1`
1. `git checkout --track upstream/release-0.14`
2. `git tag -a v0.14.1` (create new release tag)
3. `git tag -fa stable` (move `stable` tag)
4. `git push upstream --tags --force` (force needed for moving `stable` tag)
5. Draft new release in the GitHub UI **with the existing new tag** (`v0.14.1` in this case)

6. Create a new release on GitHub. The following is automated:
    - Tagging the branch on GitHub
    - Creating a change log on GitHub
    - Version number of the Python package is derived from the tag name
7. Validate the tag release (manual testing).
8. Publish the release on GitHub. The following is automated:

>     - Publishing to PyPI. Once published, a release cannot be overridden. Bad releases can be yank by a PyPI maintainer.

Reviewer 3: Normally the automation I would expect is that this would happen immediately and automatically after step 6 above (creating the tag in git). You have it listed here as after some manual testing. Either way is OK, but this would require a different github workflow configuration and set of steps for how to run the automation. I want to make sure we're clear on the intended workflow so I can help ensure it's configured accordingly.

Reviewer 4: A git tag can also be created locally with `git tag` and pushed to GitHub with `git push --tags` / `git push upstream refs/tags/tagname`. Or are you going to prevent tag pushes from users? A GH release can either use an existing tag or create a new tag on `publish`.

My #922 uses the event `github.event.action == 'published'` to push a release to PyPI and to attach the release files and sigstore signatures to the GH release. The event is fired when a privileged user clicks the green `publish` button on a GH release.

The release publish event is my favorite. We can use different approach, e.g. publish event with [deployment environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) or manual workflow dispatch. Environments can be used for additional protection, to delay release jobs, and to enforce reviews. I have never used deploy environment for any of my personal projects and I'm not familiar how they work in detail.

>     - Publishing to PyPI. Once published, a release cannot be overridden. Bad releases can be yank by a PyPI maintainer.

Reviewer 3: Are you sure PyPI publishing won't happen until this step? If the tag is created at step 4 above, I think that's when PyPI will happen automatically.

https://github.com/instructlab/instructlab/blob/main/.github/workflows/pypi.yaml

Reviewer 3: nevermind, this is wrong.  THe latest workflow that I linked to limits publishing to the release publish event

9. Announce release on the following channels:

>     - Community Slack

Reviewer 3: ```suggestion
    - InstructLab Slack
```

Reviewer 2: Public email lists as well should be included here

Author: @lhawthorn Are the email lists active yet?  I'd be happy to include them and a release page on the website on this list in a future release.

10. Create a milestone on GitHub for the next release without a milestone.

>     - The date for this milestone is set to 2-3 weeks after the next milestone. This date can be tweaked at the next refinement session.

Reviewer 2: Again, let's have timeline/schedule stuff be in one place - repeating ourselves just means more to maintain/more chances for these docs to be inconsistent

# InstructLab CLI Release Strategy

This document discusses the release strategy and processes for the
`instructlab` Python package built from the
<https://github.com/instructlab/instructlab> git repository.

## Versioning Scheme

Releases use a `X.Y.Z` numbering scheme.

X-stream release are for major releases. At this stage in the project a major release has not been cut and we expect each release (2-3 weeks) to be a new Y-stream.

> Backwards compatibility will exist within a Y-stream for example within 0.1.8 and 0.1.9 and between Y-stream releases for update purposes. For example between 0.1.9 and 0.2.0.

Author: @cdoern I should probably make the wording say subsequent Y-stream releases right? We're not ensuring compatibility between 0.1.9 and 0.3.X but we are 0.2.9 and 0.3.X?

Reviewer 3: I would take out the mention of update compatibility. I'm not totally sure what that means. I think you just mean you can `pip upgrade` or whatever? I'm not sure how to NOT allow that and I don't want anyone to think we're doing anything extra clever.

Author: @tiran Do you know if there are any expectations the python community has around `pip upgrade` or if you have any comments update compatibility in general?

Reviewer 4: No, projects are free to define their own compatibility and update policy. People often assume CalVer or SemVer these days, but nobody is forcing projects to adopt SemVer or CalVer. Also we are < 1.0, so there is no backwards compatibility promise anyway.

Reviewer 3: To @tiran's point about pre-1.0, I think I would remove the comment about compatibility of upgrades. If we say that, I think we need to get more detailed about what that means. For example, there's a proposal about completely changing the CLI subcommand layout (a good thing, I like it to be clear!) but that breaks compatibility from a user perspective. I don't think we're quite far enough along to require maintaining backwards compatibility with all of the old commands.

Z-stream releases are meant for critical bug and documentation fixes. Backward compatibility is guaranteed between Z-stream releases. Z-stream releases are cut as maintainers see fit.

## Schedule

The project currently operates on a time-based release schedule.

> Before releasing version 1.0, the goal is to release a new Y-stream every 2-3 weeks though the time span remains flexible based on the discretion of the maintainers team.

Reviewer 3: ```suggestion
The goal is to release a new Y-stream every 2-3 weeks though the time span remains flexible based on the discretion of the maintainers team.
```

I don't think it needs a qualifier about 1.0

Reviewer 2: Agreed and also don't think we need to stick to a schedule here, there's a separate section in the doc for that

The cadence for major releases starting from 1.0 and onward will be determined as the project matures.

A schedule will be updated in a markdown file on the <https://github.com/instructlab/instructlab> Github repository.

## Release Tracking

Release planning is tracked via milestones in Github. Milestones on Github will exist for the next two releases.

> Issues and PRs for milestones are set, monitored, and changed at weekly grooming sessions.

Reviewer 3: I don't think this should be here until such a thing is established in a public time and place

Reviewer 5: Also fyi industry is moving away from using the term `grooming` to use `refinement` instead

> Only PRs and issues associated with the next two milestones will be prioritized and merged into the main branch of instructlab

Reviewer 3: "Only" seems too strong here. If as a maintainer, I jump on the PR queue and see a good bug fix that was just posted, merging it should be fine, even if it hasn't gone through a grooming session. It just seems like overhead that shouldn't be strictly always required.

```suggestion 
PRs and issues associated with the next two milestones will be prioritized for review and merging into the main branch of instructlab.
```

## Git Branches and Tags

Every `X.Y` release stream gets a new branch.

> Each release, `X.Y.Z`, exists as a tag on the `X.Y` branch.

Reviewer 6: https://github.com/instructlab/instructlab/blob/bc0bdf33f1e78876e109911f535d1d89918f63e3/.github/workflows/pypi.yaml#L9

assumes tags start with `v`.

Reviewer 3: ```suggestion
Each release, `X.Y.Z`, exists as a tag named `vX.Y.Z` on the `X.Y` branch.
```

## Release Branch Maintenance

Maintenance efforts are primarily focused on the `main` branch.
Occasionally, critical bug fixes may be backported to the most recent
`X.Y` branch for a Z-stream bug-fix release. Bug fixes in general are not
actively backported to release branches.

> We will not have a X-stream release unless a total re-architecture of instructlab is occurring.

Reviewer 3: Well, I assume 1.0 would happen someday without a re-architecture.

The line could probably be removed ...

## Release Mechanics

The following are the steps for how a release gets cut.

1. Maintainers determine a commit on the main branch that will serve as the basis for the next release.
    - For a Z-stream release skip this step.
2. For a y-Stream release: create a new branch in the format release-X.YY.0.
    - For a Z-Stream release skip this step

> 3. Create a commit for the release

Reviewer 3: What would the contents of this commit include?

> 3. Create a commit for the release

Reviewer 2: Why do we need a commit at all?

4. Open up a PR
5. Maintainers Merge PR

>     - The PR will require 2 maintainer approvals.

Reviewer 3: What would the PR include?

> 7. Create a new release on Github. The following is automated
>     - Publishing to PyPI
>     - Tagging the branch on Github
>     - Creating a change log on Github

Reviewer 4: Publishing to PyPI happens when the release is published, not when it is created. This allows us to run some checks on a release build before it gets pushed to PyPI. PyPI does not permit users to override files. A PyPI maintainer can yank a release, but they cannot override a release.

```suggestion
7. Create a new release on Github. The following is automated
    - Tagging the branch on Github
    - Creating a change log on Github
8. Validate the tag release (manual testing)
9. Publish the release on Github. The following is automated
    - Publishing to PyPI. Once published, a release cannot be overridden. Bad releases can be yank by a PyPI maintainer.
```

Reviewer 4: The document does not define how a release tag should look like. We currently prefix the release tag with `v`, so it's `vX.Y.Z`. That's a recommended schema for setuptools-scm, too.

Reviewer 7: @tiran is the creation and publishing of releases a definitively separate process? I'm not sure, I haven't published a release on GH before.

If so, I agree with your suggestion.

8. Announce release on the following channels
    - Community Slack

>     - The Instructlab community mailing list

Reviewer 3: I would drop this since it doesn't exist (yet, anyway)

>     - Update the releases page on the Instructlab website

Reviewer 3: This doesn't exist either, right?
